### ISSUE TITLE
Allowed changes in DomainEventHandler - lack of example

### ISSUE THREAD

**Issue by rmaziarka** [kgrzybek/sample-dotnet-core-cqrs-api#1]
Hello Kamil,

Thanks for that repo - a lot of valuable knowledge 😊

I have a question regarding allowed changes in your DomainEventHandlers - there is no code in your [OrderAddedDomainEventHandler.cs](https://github.com/kgrzybek/sample-dotnet-core-cqrs-api/blob/master/src/SampleProject.API/Orders/AddCustomerOrder/OrderAddedDomainEventHandler.cs). It seems that changes in another aggregate in handler won't be published as events. For example:

- In OrderAddedDomainEventHandler we get Buyer aggregate
- We add order to the buyer
- We have an event to publish but nothing will gather this event
    - If we run CommitAsync from context code will SaveChangesAsync twice
    - If not we won't run DispatchEventsAsync from the dispatcher

So we are not able to implement fully the scenario of triggering side effects across multiple aggregates ([Microsoft docs](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/domain-events-design-implementation#domain-events-as-a-preferred-way-to-trigger-side-effects-across-multiple-aggregates-within-the-same-domain))

What do you think of it? Maybe you have a different perspective on this topic 😉

**Comment by kgrzybek** [comment]
Hi Radek,

Thanks for feedback!

You are definitely right. On the beginning of this repo I didn't focus on more advanced scenarios when one command can cause the flood of many events from various aggregates (and handling issues in the same transaction boundary what is not recommended but sometimes necessary OR in separate transaction = eventual consistency).

Because DDD is just starting to shine in such complex situations, I decided to fix my error quickly and commited new changes :)

The main change is an introduction of Autofac Decorators which are responsible for dispatching events and committing transaction:
[DomainEventsDispatcherCommandHandlerDecorator.cs](https://github.com/kgrzybek/sample-dotnet-core-cqrs-api/blob/master/src/SampleProject.Infrastructure/DomainEventsDispatcherCommandHandlerDecorator.cs)
[DomainEventsDispatcherNotificationHandlerDecorator.cs](https://github.com/kgrzybek/sample-dotnet-core-cqrs-api/blob/master/src/SampleProject.Infrastructure/DomainEventsDispatcherNotificationHandlerDecorator.cs)

As you see the former is dispatching events and committing transaction and the latter only is dispatching events to avoid multiple commits.

I have added more advanced scenarios:
AddOrderCommand > AddOrderEvent->new Payment Aggregate created (same transaction) -> PaymentCreatedEvent -> Saving to outbox -> PaymentCreatedNotificationHandler (sending email)

RegisterCustomerCommand -> CustomerRegisteredEvent (end of transaction)-> Saving to outbox -> CustomerRegisteredNotificationHandler sends email and invokes aggregate method on Customer Aggregate (separate transaction - eventual consistency).

Let me know how you see actual implementation. I had to introduce small changes as well (for example separate UoW class) but Domain code is almost intact (only new things).

PS: I will definitely write another post on my blog about it :)
